Fix raylet pinning race condition

edoakes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?

There's currently a race condition where an object could be evicted between when it's created and when the raylet receives the `PinObjectIDs` RPC and gets the buffer. This change fixes it by delaying the plasma `Release` call until after the RPC response is received. Also modifies the `CoreWorker::Put` and `CoreWorker::Seal` interfaces a bit so that `PinObjectIDs` is only called there instead of in a few different places in the code.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

edoakes commented on src/ray/core_worker/transport/direct_actor_transport.cc:
```diff
@@ -284,15 +282,6 @@ void CoreWorkerDirectTaskReceiver::HandlePushTask(
-      // here and when raylet pins them.
-      if (!plasma_return_ids.empty()) {
-        RAY_CHECK_OK(
-            local_raylet_client_->PinObjectIDs(caller_address, plasma_return_ids));
```
Note that the change to do this in `core_worker.cc` means we're now making one RPC per spilled return ID. IMO this is not a concern but I can change it back if need be - it just makes this already complicated logic messier.

stephanie-wang commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -400,20 +391,31 @@ Status CoreWorker::Put(const RayObject &object,
+          }
+        }));
+  } else {
+    RAY_RETURN_NOT_OK(plasma_store_provider_->Release(object_id));
```
Should we do this if object_exists is true? Wasn't sure about the semantics of plasma's `Put`.

edoakes replied:
Oh that's a great catch. We should not - updating.

edoakes commented:
Confirmed that we no longer see the relevant error messages from the raylet when running the reference counting tests.

@stephanie-wang do you think we should make it a `RAY_CHECK` in the raylet RPC handler that the object was available?

edoakes commented:
Actually I don't think that's safe - the worker could crash before the raylet processes the message.

stephanie-wang commented on src/ray/core_worker/transport/direct_actor_transport.cc:
```diff
@@ -284,15 +282,6 @@ void CoreWorkerDirectTaskReceiver::HandlePushTask(
-      // lifetimes.
-      // TODO(edoakes): the plasma objects could be evicted between creating them
-      // here and when raylet pins them.
-      if (!plasma_return_ids.empty()) {
```
I think we can also remove `plasma_return_ids` now.

stephanie-wang commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -400,20 +391,31 @@ Status CoreWorker::Put(const RayObject &object,
+  bool object_exists;
+  RAY_RETURN_NOT_OK(plasma_store_provider_->Put(object, object_id, &object_exists));
+  if (pin_object && !object_exists) {
+    // Tell the raylet to pin the object **after** it is created.
```
This block seems complicated enough to move out into a separate method.

edoakes replied:
I considered that, but there's actually not much happening here - just a single RPC call with a callback that releases the object - and moving it to a separate method would require either splitting up the release logic in sort of an ugly way or passing in a callback to that method in which case it would basically look the same.

stephanie-wang replied:
Oh I was thinking of moving out the whole if-else block (everything except `!object_exists`), but not a big deal.

stephanie-wang commented on src/ray/core_worker/store_provider/plasma_store_provider.h:
```diff
@@ -37,11 +40,35 @@ class CoreWorkerPlasmaStoreProvider {
+  ///
+  /// \param[in] metadata The metadata of the object.
+  /// \param[in] data_size The size of the object.
+  /// argument to Get to retrieve the object data.
```
Remove?

stephanie-wang commented on src/ray/core_worker/store_provider/plasma_store_provider.h:
```diff
@@ -37,11 +40,35 @@ class CoreWorkerPlasmaStoreProvider {
+  /// \param[in] data_size The size of the object.
+  /// argument to Get to retrieve the object data.
+  /// \param[in] object_id The ID of the object. This can be used as an
+  /// argument to Get to retrieve the object data.
```
Remove?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22153/
Test FAILed.

stephanie-wang commented on src/ray/core_worker/core_worker.h:
```diff
@@ -239,7 +240,8 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
   /// \return Status.
-  Status Seal(const ObjectID &object_id, bool pin_object);
+  Status Seal(const ObjectID &object_id, bool pin_object,
+              absl::optional<rpc::Address> owner_address = absl::nullopt);
```
Document the arg.

stephanie-wang replied:
```suggestion
              const absl::optional<rpc::Address> &owner_address = absl::nullopt);
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22152/
Test FAILed.

stephanie-wang commented:
Oh yeah, I don't think it's safe to make it an assertion. Maybe update the error message to say that it should only happen because of a worker failure, though.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22154/
Test FAILed.

edoakes commented:
@stephanie-wang addressed the comments.

Yep, that test is skipped in https://github.com/ray-project/ray/pull/7218 but tested locally on this PR and didn't see the issue. Will merge that PR first then uncomment here.

stephanie-wang commented on src/ray/core_worker/core_worker.h:
```diff
@@ -239,9 +239,11 @@ class CoreWorker : public rpc::CoreWorkerServiceHandler {
   ///
   /// \param[in] object_id Object ID corresponding to the object.
   /// \param[in] pin_object Whether or not to pin the object at the local raylet.
+  /// \param[in] owner_address Address of the owner of the object who will be contacted
```
```suggestion
  /// \param[in] owner_address Address of the owner of the object who will be contacted by the raylet
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22155/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22156/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22161/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22158/
Test FAILed.
